[BUG] mastra output exceeds the cloudflare size limit due to bundling typescript

Thank you for creating this great project ☺️

### Describe the Bug

mastra build output depends `typescript` and wrangler bundle it when deploying. 

```
grep -r "'typescript'" .mastra/output/dist-*
.mastra/output/dist-MEN73GGI.mjs:      const ts = await import('typescript');
.mastra/output/dist-MEN73GGI.mjs:      const ts = await import('typescript');
.mastra/output/dist-MEN73GGI.mjs:      const ts = await import('typescript');
```

Therefor, Cloudflare deployment fails.

> Your Worker exceeded the size limit of 3 MiB.


```
$ pnpm run build && pnpm run deploy

***

INFO [2025-12-27 11:46:12.652 +0900] (Mastra CLI): Deployer found, preparing deployer build...
INFO [2025-12-27 11:46:12.843 +0900] (Mastra CLI): Analyzing dependencies...
INFO [2025-12-27 11:46:13.483 +0900] (Mastra CLI): Optimizing dependencies...
INFO [2025-12-27 11:46:19.375 +0900] (Mastra CLI): Bundling Mastra application
INFO [2025-12-27 11:46:21.442 +0900] (Mastra CLI): Bundling Mastra done
INFO [2025-12-27 11:46:21.442 +0900] (Mastra CLI): Copying public files
INFO [2025-12-27 11:46:21.442 +0900] (Mastra CLI): Done copying public files
INFO [2025-12-27 11:46:21.442 +0900] (Mastra CLI): Copying .npmrc file
INFO [2025-12-27 11:46:21.442 +0900] (Mastra CLI): Done copying .npmrc file
INFO [2025-12-27 11:46:21.442 +0900] (Mastra CLI): Installing dependencies
INFO [2025-12-27 11:46:25.090 +0900] (Mastra CLI): Done installing dependencies
INFO [2025-12-27 11:46:25.090 +0900] (Mastra CLI): You can now deploy the .mastra/output directory to your target platform.

***


 ⛅️ wrangler 4.54.0
───────────────────
Total Upload: 19010.01 KiB / gzip: 3250.35 KiB
Your Worker has access to the following bindings:
Binding                          Resource                  
env.NODE_ENV ("production")      Environment Variable      

✘ [ERROR] Your Worker failed validation because it exceeded size limits.

  
  A request to the Cloudflare API
  (/accounts/***/workers/scripts/mastra-cloudflare-deployer/versions)
  failed.
   - Your Worker exceeded the size limit of 3 MiB. Please upgrade to a paid plan to deploy Workers
  up to 10 MiB. https://dash.cloudflare.com/***/workers/plans [code:
  10027]
  To learn more about this error, visit:
  https://developers.cloudflare.com/workers/platform/limits/#worker-size
  Here are the 5 largest dependencies included in your script:
  
  - node_modules/.pnpm/typescript@5.9.3/node_modules/typescript/lib/typescript.js - 10788.92 KiB
  - mastra.mjs - 5571.02 KiB
  - dist-MEN73GGI.mjs - 1887.02 KiB
  - index3.mjs - 52.81 KiB
  - dist-4CMHRWC4.mjs - 41.06 KiB
  
  If these are unnecessary, consider removing them
  



✘ [ERROR] A request to the Cloudflare API (/accounts/***/workers/scripts/mastra-cloudflare-deployer/versions) failed.

  Your Worker exceeded the size limit of 3 MiB. Please upgrade to a paid plan to deploy Workers up
  to 10 MiB. https://dash.cloudflare.com/***/workers/plans [code:
  10027]
  To learn more about this error, visit:
  https://developers.cloudflare.com/workers/platform/limits/#worker-size
```

### Steps To Reproduce

1. git clone https://github.com/TatsuyaYamamoto/mastra-cloudflare-deployer.git
2. `pnpm run build && pnpm run deploy`

### Link to Minimal Reproducible Example

https://github.com/TatsuyaYamamoto/mastra-cloudflare-deployer

### Expected Behavior

For Cloudflare deployment, the goal is to remove the TypeScript dependency, but it’s unclear whether this is essential for Mastra.

### Environment Information

```block
System:
    OS: macOS 15.7.1
    CPU: (12) arm64 Apple M4 Pro
    Memory: 205.25 MB / 48.00 GB
    Shell: 5.9 - /bin/zsh
  Binaries:
    Node: 24.12.0
    npm: 11.6.2
    pnpm: 10.26.2
    bun: 1.2.21 - /opt/homebrew/bin/bun
    Deno: 2.4.5 - /opt/homebrew/bin/deno
  Browsers:
    Chrome: 143.0.7499.170
    Safari: 26.0.1
  npmPackages:
    @mastra/core: 1.0.0-beta.18 => 1.0.0-beta.18 
    @mastra/deployer-cloudflare: 1.0.0-beta.1 => 1.0.0-beta.1 
    @mastra/evals: 1.0.0-beta.3 => 1.0.0-beta.3 
    @mastra/libsql: 1.0.0-beta.10 => 1.0.0-beta.10 
    @mastra/loggers: 1.0.0-beta.3 => 1.0.0-beta.3 
    @mastra/memory: 1.0.0-beta.9 => 1.0.0-beta.9 
    @mastra/observability: 1.0.0-beta.9 => 1.0.0-beta.9 
    @types/node: ^25.0.3 => 25.0.3 
    mastra: 1.0.0-beta.12 => 1.0.0-beta.12 
    typescript: ^5.9.3 => 5.9.3 
    wrangler: 4.54.0 => 4.54.0 
    zod: ^4.2.1 => 4.2.1 
  npmGlobalPackages:
    corepack: 0.34.5
    npm: 11.6.2
```

### Verification

- [x] I have searched the existing issues to make sure this is not a duplicate
- [x] I have included sufficient information for the team to reproduce and understand the issue

--- a/.changeset/few-lemons-stay.md
+++ b/.changeset/few-lemons-stay.md
@@ -1,0 +1,5 @@
+---
+'@mastra/deployer-cloudflare': patch
+---
+
+Fixed Cloudflare Workers exceeding the 3MB size limit due to TypeScript being bundled. The deployer now stubs out TypeScript (~10MB) since the agent-builder gracefully falls back to basic validation when it's unavailable.

--- a/deployers/cloudflare/src/index.test.ts
+++ b/deployers/cloudflare/src/index.test.ts
@@ -1,0 +1,68 @@
+import { mkdir, readFile, rm } from 'node:fs/promises';
+import { tmpdir } from 'node:os';
+import { join } from 'node:path';
+import { describe, it, expect, beforeEach, afterEach, vi } from 'vitest';
+
+import { CloudflareDeployer } from './index.js';
+
+describe('CloudflareDeployer', () => {
+  let deployer: CloudflareDeployer;
+  let tempDir: string;
+
+  beforeEach(async () => {
+    tempDir = join(tmpdir(), `cloudflare-deployer-test-${Date.now()}`);
+    // Create the output directory that writeFiles expects
+    await mkdir(join(tempDir, 'output'), { recursive: true });
+  });
+
+  afterEach(async () => {
+    await rm(tempDir, { recursive: true, force: true });
+  });
+
+  describe('writeFiles', () => {
+    describe('TypeScript stub for bundle size optimization', () => {
+      it('should create typescript-stub.mjs and configure wrangler alias', async () => {
+        deployer = new CloudflareDeployer({ name: 'test-worker' });
+        vi.spyOn(deployer, 'loadEnvVars').mockResolvedValue(new Map());
+
+        await deployer.writeFiles(tempDir);
+
+        // Verify stub file is created with expected exports
+        const stubPath = join(tempDir, 'output', 'typescript-stub.mjs');
+        const stub = await import(stubPath);
+
+        expect(stub.default).toEqual({});
+        expect(stub.createSourceFile()).toBeNull();
+        expect(stub.createProgram()).toBeNull();
+        expect(stub.ScriptTarget.Latest).toBe(99);
+        expect(stub.DiagnosticCategory.Error).toBe(1);
+
+        // Verify wrangler config uses the stub
+        const wranglerConfigPath = join(tempDir, 'output', 'wrangler.json');
+        const wranglerConfig = JSON.parse(await readFile(wranglerConfigPath, 'utf-8'));
+
+        expect(wranglerConfig.alias.typescript).toBe('./typescript-stub.mjs');
+      });
+
+      it('should allow user to override the TypeScript alias', async () => {
+        deployer = new CloudflareDeployer({
+          name: 'test-worker',
+          alias: {
+            typescript: './custom-typescript-stub.js',
+            'other-module': './other.js',
+          },
+        });
+        vi.spyOn(deployer, 'loadEnvVars').mockResolvedValue(new Map());
+
+        await deployer.writeFiles(tempDir);
+
+        const wranglerConfigPath = join(tempDir, 'output', 'wrangler.json');
+        const wranglerConfig = JSON.parse(await readFile(wranglerConfigPath, 'utf-8'));
+
+        // User's alias should override the default, other aliases preserved
+        expect(wranglerConfig.alias.typescript).toBe('./custom-typescript-stub.js');
+        expect(wranglerConfig.alias['other-module']).toBe('./other.js');
+      });
+    });
+  });
+});

--- a/deployers/cloudflare/src/index.ts
+++ b/deployers/cloudflare/src/index.ts
@@ -61,11 +61,37 @@ export class CloudflareDeployer extends Deployer {
   }
 
   async writeFiles(outputDirectory: string): Promise<void> {
-    const { vars: userVars, ...userConfig } = this.userConfig;
+    const { vars: userVars, alias: userAlias, ...userConfig } = this.userConfig;
     const loadedEnvVars = await this.loadEnvVars();
 
     // Merge env vars from .env files with user-provided vars
     const envsAsObject = Object.assign({}, Object.fromEntries(loadedEnvVars.entries()), userVars);
+
+    // Write TypeScript stub to prevent bundling the full TypeScript library (~10MB)
+    // The agent-builder package dynamically imports TypeScript for code validation,
+    // but gracefully falls back to basic validation when TypeScript is unavailable.
+    // This stub ensures the import doesn't fail while keeping the bundle small.
+    const typescriptStubPath = 'typescript-stub.mjs';
+    const typescriptStub = `// Stub for TypeScript - not available at runtime in Cloudflare Workers
+// The @mastra/agent-builder package will fall back to basic validation
+export default {};
+export const createSourceFile = () => null;
+export const createProgram = () => null;
+export const findConfigFile = () => null;
+export const readConfigFile = () => ({ error: new Error('TypeScript not available') });
+export const parseJsonConfigFileContent = () => ({ errors: [new Error('TypeScript not available')], fileNames: [], options: {} });
+export const flattenDiagnosticMessageText = (message) => typeof message === 'string' ? message : message?.messageText || '';
+export const ScriptTarget = { Latest: 99 };
+export const ModuleKind = { ESNext: 99 };
+export const JsxEmit = { ReactJSX: 4 };
+export const DiagnosticCategory = { Warning: 0, Error: 1, Suggestion: 2, Message: 3 };
+export const sys = {
+  fileExists: () => false,
+  readFile: () => undefined,
+};
+`;
+
+    await writeFile(join(outputDirectory, this.outputDir, typescriptStubPath), typescriptStub);
 
     const wranglerConfig: Unstable_RawConfig = {
       name: 'mastra',
@@ -79,6 +105,11 @@ export class CloudflareDeployer extends Deployer {
       ...userConfig,
       main: './index.mjs',
       vars: envsAsObject,
+      // Alias TypeScript to stub to prevent wrangler from bundling the full library
+      alias: {
+        typescript: `./${typescriptStubPath}`,
+        ...userAlias,
+      },
     };
 
     await writeFile(join(outputDirectory, this.outputDir, 'wrangler.json'), JSON.stringify(wranglerConfig));

